sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala

```diff
@@ -597,28 +602,29 @@ class GpuDynamicPartitionDataSingleWriter(
    */
   private def writeTableAndClose(
     table: Table, outDataTypes: Array[DataType], maxRecordsPerFile: Long, partPath: String) = {
-    val splitIndexes = getSplitIndexes(
-      maxRecordsPerFile,
-      currentWriterStatus.recordsInFile,
-      table.getRowCount()
-    )
-    var needNewWriter = currentWriterStatus.recordsInFile >= maxRecordsPerFile
-
     val tabs = withResource(table) { _ =>
+      assertNotExceedMaxFileCounter(
+        currentWriterStatus.fileCounter,
+        maxRecordsPerFile,
+        currentWriterStatus.recordsInFile,
+        table.getRowCount(),
+        MAX_FILE_COUNTER
+      )
+      val splitIndexes = getSplitIndexes(
+        maxRecordsPerFile,
+        currentWriterStatus.recordsInFile,
+        table.getRowCount()
+      )
       table.contiguousSplit(splitIndexes: _*)
     }
     withResource(tabs) { _ =>
+      var needNewWriter = currentWriterStatus.recordsInFile >= maxRecordsPerFile
       tabs.foreach(b => {
         if (needNewWriter) {
           currentWriterStatus.fileCounter += 1
-          assert(currentWriterStatus.fileCounter <= MAX_FILE_COUNTER,
-            s"File counter ${currentWriterStatus.fileCounter} " +
-              s"is beyond max value $MAX_FILE_COUNTER")
 
           // will create a new file, close the old writer
-          if (currentWriterStatus != null) {
-            releaseWriter(currentWriterStatus.outputWriter)
-          }
+          releaseWriter(currentWriterStatus.outputWriter)
```

Comment:
Better to not remove the `if` check, this can avoid the NPE when accessing the `currentWriterStatus.outputWriter`.

Reply:
`currentWriterStatus` has been accessed multiply times above this line of code. So we can make sure there will be no `NPE` thrown from this line of code.